### PR TITLE
`RUF031`: Ignore unparenthesized tuples in subscripts when the subscript is obviously a type annotation or type alias

### LINKED ISSUE #12758: RUF031 with `parenthesize-tuple-in-subscript = true` should ignore type annotations
Feels like a bug to me that this applies to type annotations. In case of dict keys it's a very nice rule to make it clear that you have a tuple key, but for type annotations it feels pretty noisy/ugly...

```py
$ ruff version
ruff 0.5.7

$ cat test.py
bar = {}
bar[('hello', 'world')] = None
foo: tuple[str, str] | None = None

$ ruff check --isolated --select RUF031 --config 'lint.ruff.parenthesize-tuple-in-subscript = true' --preview --no-cache test.py
test.py:3:12: RUF031 [*] Use parentheses for tuples in subscripts.
  |
1 | bar = {}
2 | bar[('hello', 'world')] = None
3 | foo: tuple[str, str] | None = None
  |            ^^^^^^^^ RUF031
  |
  = help: Parenthesize the tuple.

Found 1 error.
[*] 1 fixable with the `--fix` option.
``

### PR DESCRIPTION
**Pull request by AlexWaygood**
## Summary

Fixes #12758. Previously with the setting `lint.ruff.parenthesize-tuple-in-subscript = true`, RUF031 would have complained about annotations such as `dict[str, int]`, telling you that you should spell them as `dict[(str, int)]` instead. That feels somewhat silly -- although it is of course a subscription in exactly the same way as a dictionary key lookup, it's not really what the rule is trying to detect.

This PR fixes most of those cases. Unfortunately it doesn't fix cases like this, where it's ambiguous whether the variable is a type alias or not:

```py
import typing
X = typing.Callable[[str, int], bool]  # RUF031 here
```

Faced with this, however, users have a fairly simple workaround: make it explicit that it's a type alias, and the false positive will go away

```py
import typing
X: typing.TypeAlias = typing.Callable[[str, int], bool]
```

## Test Plan

`cargo test -p ruff_linter --lib`

### CODE REVIEW

**Comment by AlexWaygood** [comment]
> [RasaHQ/rasa](https://github.com/RasaHQ/rasa) (+0 -1 violations, +0 -0 fixes)
> 
> ruff check --no-cache --exit-zero --ignore RUF9 --output-format concise --preview
> 
> - [tests/utils/test_train_utils.py:117:67:](https://github.com/RasaHQ/rasa/blob/7807b19ad5fffab73ca1a04dc710f812115a9288/tests/utils/test_train_utils.py#L117) RUF031 [*] Avoid parentheses for tuples in subscripts.

Oh, that's ugly. Sorta feels like a shame that we'll no longer be emitting a diagnostic there 😆

I suppose we *could* only ignore the rule for type annotations/aliases if `lint.ruff.parenthesize-tuple-in-subscript = true`, and continue to apply the rule everywhere if `lint.ruff.parenthesize-tuple-in-subscript = false`? But that feels conceptually harder to explain

**Comment by dylwil3** [comment]
Does it make sense to say in the docstring for the rule that we skip annotations, or is that too in the weeds?

**Comment by AlexWaygood** [comment]
That's a good idea. I'll add some more docs tomorrow (it's late now where I am :-)

File: crates/ruff_linter/src/rules/ruff/rules/incorrectly_parenthesized_tuple_in_subscript.rs [hunk 1]
```diff
@@ -64,6 +64,10 @@ pub(crate) fn subscript_with_parenthesized_tuple(checker: &mut Checker, subscrip
     if tuple_subscript.parenthesized == prefer_parentheses || tuple_subscript.elts.is_empty() {
         return;
     }
+    let semantic = checker.semantic();
+    if semantic.in_annotation() || semantic.in_type_definition() {
```
Comment (dylwil3) [inline crates/ruff_linter/src/rules/ruff/rules/incorrectly_parenthesized_tuple_in_subscript.rs:68]: Should we also do `semantic.in_simple_string_type_definition()` here?

**Comment by dhruvmanila** [comment]
I think it might make sense to cover generics as well (https://github.com/astral-sh/ruff/issues/12773#issuecomment-2277195957) like:

```py
class Foo(Generic[T1, T2]):
	...
```

(It can be done as a follow-up.)

**Comment by MichaReiser** [comment]
> such as dict[str, int], telling you that you should spell them as dict[(str, int)] instead. That feels somewhat silly -- although it is of course a subscription in exactly the same way as a dictionary key lookup, it's not really what the rule is trying to detect.

I find it hard to following this reasoning but it's probably just me not writing enough python. Isn't this as "silly" as doing the same in regular subscripts?

**Comment by AlexWaygood** [comment]
> I find it hard to following this reasoning but it's probably just me not writing enough python. Isn't this as "silly" as doing the same in regular subscripts?

Type annotations just have their own specific conventions around them. And the subscription has a very different meaning in a typing context. You're not "looking up" a value at an index or key in a typing context in the same way; and I think the vast majority of people aren't even aware that they're creating implicit tuples in a type annotation context.

File: crates/ruff_linter/src/rules/ruff/rules/incorrectly_parenthesized_tuple_in_subscript.rs (hunk 1 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/ruff/rules/incorrectly_parenthesized_tuple_in_subscript.rs]: It's not needed -- if the `SIMPLE_STRING_TYPE_DEFINITION` flag has been set on the semantic model, the `TYPE_DEFINITION` flag will also have been set:

https://github.com/astral-sh/ruff/blob/c4e651921ba1911a04102e2052c9fe56f185309d/crates/ruff_linter/src/checkers/ast/mod.rs#L2196-L2204
